python/mxnet/rnn/rnn_cell.py

```diff
@@ -0,0 +1,503 @@
+    def output_shape(self):
+        """shape(s) of output"""
+        raise NotImplementedError()
+
+    def begin_state(self, init_sym=symbol.zeros, **kwargs):
+        """Initial state for this cell.
+
+        Parameters
+        ----------
+        init_sym : Symbol, default symbol.zeros
+            Symbol for generating initial state. Can be zeros,
+            ones, uniform, normal, etc.
+        **kwargs :
+            more keyword arguments passed to init_sym. For example
+            mean, std, dtype, etc.
+
+        Returns
+        -------
+        states : nested list of Symbol
+            starting states for first RNN step
+        """
+        assert not self._modified, \
+            "After applying modifier cells (e.g. DropoutCell) the base " \
+            "cell cannot be called directly. Call the modifier cell instead."
+        state_shape = self.state_shape
+        def recursive(shape):
+            """Recursively construct input states"""
+            if isinstance(shape, tuple):
+                assert len(shape) == 0 or isinstance(shape[0], numeric_types)
+                self._init_counter += 1
+                return init_sym(name='%sinit_%d'%(self._prefix, self._init_counter),
+                                shape=shape, **kwargs)
+            else:
+                assert isinstance(shape, list)
+                return [recursive(i) for i in shape]
+
+        return recursive(state_shape)
+
+    #pylint: disable=no-self-use
+    def _get_activation(self, inputs, activation, **kwargs):
+        """Get activation function. Convert if is string"""
+        if isinstance(activation, string_types):
+            return symbol.Activation(inputs, act_type=activation, **kwargs)
+        else:
+            return activation(inputs, **kwargs)
+
+
+class RNNCell(BaseRNNCell):
+    """Simple recurrent neural network cell
+
+    Parameters
+    ----------
+    num_hidden : int
+        number of units in output symbol
+    activation : str or Symbol, default 'tanh'
+        type of activation function
+    prefix : str, default 'rnn_'
+        prefix for name of layers
+        (and name of weight if params is None)
+    params : RNNParams or None
+        container for weight sharing between cells.
+        created if None.
+    """
+    def __init__(self, num_hidden, activation='tanh', prefix='rnn_', params=None):
+        super(RNNCell, self).__init__(prefix=prefix, params=params)
+        self._num_hidden = num_hidden
+        self._activation = activation
+        self._iW = self.params.get('i2h_weight')
+        self._iB = self.params.get('i2h_bias')
+        self._hW = self.params.get('h2h_weight')
+        self._hB = self.params.get('h2h_bias')
+
+    @property
+    def state_shape(self):
+        """shape(s) of states"""
+        return (0, self._num_hidden)
+
+    @property
+    def output_shape(self):
+        """shape(s) of output"""
+        return (0, self._num_hidden)
+
+    def __call__(self, inputs, states):
```

Comment:
@piiswrong @pluskid Should we also add a "seq_len" or "input_length" argument to this \_\_call__ function? Like the "input_length" in Keras https://github.com/fchollet/keras/blob/master/keras/layers/recurrent.py#L115-L123.